File: 백준/Gold/20183. 골목 대장 호석 － 효율성 2/골목 대장 호석 － 효율성 2.py
import sys
import heapq

input = sys.stdin.readline

n, m, start, last, value = map(int, input().split())
graph = [[] for _ in range(n + 1)]
cc=[]
right=-1
for _ in range(m):
    a1, a2, a3 = map(int, input().split())
    heapq.heappush(cc,-a3)
    graph[a1].append((a2, a3))
    graph[a2].append((a1, a3))




def dijk(start, last, tt):

    q = []
    distance = [sys.maxsize] * (n + 1)
    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        val, ver = heapq.heappop(q)
        if distance[ver] < val:
            continue
        for go, va in graph[ver]:
            if va > tt:
                continue
            if va + val <= value:
                if distance[go] > va + val:
                    distance[go] = va + val
                    if go != last:
                        heapq.heappush(q, (va + val, go))
    if distance[last]<=value and distance[last]!=-1:
        return 1
    else:
        return 0

left = 0
right=(-heapq.heappop(cc))+1
ans = sys.maxsize
while left + 1 < right:

    mid = (left + right) // 2
    aa = dijk(start, last, mid)
    if aa ==1:
        right = mid
        ans = min(ans, mid)
    else:
        left = mid

if ans == sys.maxsize:
    print(-1)
else:
    print(ans)

File: 백준/Gold/20183. 골목 대장 호석 － 효율성 2/골목 대장 호석 － 효율성 2.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {


    static int n, m;

    static class go {
        int vertex;
        long value;

        go(int vertex, long value) {
            this.vertex = vertex;
            this.value = value;
        }
    }

    static List<List<go>> graph = new ArrayList<>();

    public static int dijk(int start, int fin, long cost, long ma) {
        //cost 가 최대 비용
        //ma가 하나의  edgd당 갈수있는지

        long distnace[] = new long[n + 1];
        for (int i = 0; i <= n; i++) {
            distnace[i] = Integer.MAX_VALUE;
        }
        PriorityQueue<go> q = new PriorityQueue<>((x, y) -> {
            return Long.compare(x.value, y.value);
        });
        q.add(new go(start, 0));
        distnace[start] = 0;
        while (!q.isEmpty()) {
            go a = q.poll();
            if(a.value>distnace[a.vertex])
                continue;
            for (go now : graph.get(a.vertex)) {
                long temp = a.value + now.value;
                if (temp > distnace[now.vertex] || now.value > ma || temp > cost)
                    continue;
                distnace[now.vertex] = temp;
                q.add(new go(now.vertex, temp));
                if (now.vertex == fin) {
                    return 1;
                }
            }
        }
        return -1;

    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int a, b;
        long c;
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        a = Integer.parseInt(st.nextToken());
        b = Integer.parseInt(st.nextToken());
        c = Long.parseLong(st.nextToken());
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        long right = 0;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            graph.get(from).add(new go(to, weight));
            graph.get(to).add(new go(from, weight));
            right = Math.max(right, weight);
        }
        long left = -1;
        long answer = Long.MAX_VALUE;
        right += 1;
        while (left + 1 < right) {
            long mid = (left + right) / 2;
            int temp = dijk(a, b, c, mid);
            if (temp == 1) {
                right = mid;
                answer = mid;
            } else
                left = mid;
        }
        if (answer == Long.MAX_VALUE)
            System.out.println(-1);
        else
            System.out.println(answer);


    }


}
